### ISSUE TITLE
[enhancement] Add commands to cancel previous work

### ISSUE THREAD

**Issue by krOoze** [KhronosGroup/Vulkan-Docs#1468]
There are usecases where it would be beneficial to cancel previous stale workload in the queue. That could improve latency, responsiveness, and save some watts.

I imagine it could be something like a `vkQueueCancelWork( VkQueue queue )`, which would be the same as `vkQueueWaitIdle` except it does not care if the work in the queue was executed. If this could be done at Fence granularity, then even better.

Is this something that would be feasible on some GPUs?

**Comment by TomOlson** [comment]
Might be feasible on some hardware, would certainly be very implementation dependent. It's hard to predict in the abstract how much work you could save this way - do you have specific use cases in mind, or do you know of apps that look like they would benefit? Easier to quantify the benefit if we have a concrete example.

**Comment by krOoze** [comment]
Right, I have a multiple use-cases in mind that theoretically feel like a minor improvement.  Though it might border with what I would consider a good craftsmanship, rather than what might be conciously detectable by an undiscerning user.

Firstly I should say it triggers OCD a bit when the API forces user to use a semantically wrong command. I.e. it is weird to say "finish all this work", when I actually mean "drop everything, I do not care about this workload anymore". Could potentially be fine if the cancel commands would be routed to `vk*WaitIdle` if not implementable on the platform, in the same vein that layouts and renderpasses might not do anything special on some platforms.

Anyway, some potential use-cases:

- Minimized app and shut-down should not take more than 1 ms on modern computers with their unreal perf. The app should stop using all resources as soon as it is aware of these events. Though real-time app might have two or so frames queued, and non-real-time even more. That might take some time if insisting to finish that workload.
- Similarly, a resize could make frames stale and not worth finishing. Cancelling workload could prevent a small hitch at the start of sizing, and could help dedicate resources to the sizing itself (such as creating the new swapchain and producing mock frames in a timely manner).
- In non-real app it seems sane to buffer workload on the GPU little bit ahead. Though that poses a problem if e.g. the user wishes to cancel that work and do something else instead, because the work might take lets say 2 s to finish.
- In distributed compute work (including e.g. cryptocurrency mining) a work might become stale. Insisting on that work to finish in full is a small waste of watts and time\latency.
- I think there might be ways to abuse this to help real-time apps in small ways. Lot is done to preserve the phi effect (i.e. producing intermediate frames), but say I know there will be a discrete change, I could skip all that. E.g. if I know user intends to open menu or change camera dramatically etc., the queued work could be dropped and 100 % perf dedicated to rendering the new state. I.e. it should be ok if the previous frame is left little bit older as long as it helps latency of the new frame that is not related to the previous frame in temporal fashion anyway.

**Comment by msharov** [comment]
I would add another use case: WebGPU and similar projects. It is possible for a shader to hang in an infinite loop, and it is not possible to detect whether this would happen just by looking at the shader. Any application that uses shaders accepted from another sources would be vulnerable to a DoS attack, since without the ability to cancel work in progress, the only remedy is a hard reboot.

**Comment by krOoze** [comment]
@msharov Well, it should result in a Lost Device. Though it would also be nice to expose and parametrize the watchdog. On that front it might be interesting what they will do with Safety Critical version...

**Comment by HiWiSciFi** [comment]
Another use case to add to the pile:
A Mandelbrot renderer that does a multistage calculation starting off with a low resolution when moving around and increasing the resolution if the user stays on the same spot for multiple frames. High iteration counts and high resolution can easily lead to very long compute shader execution times. If the user drags the camera around while a high resolution image is being calculated then all the output data of that shader will be out-of-date which may already be known right after it was dispatched.

--> Being able to cancel the compute queue would be beneficial especially for weaker gpus since that would lead to less computational power being used for calculating data that will be discarded anyway

**Comment by pc-john** [comment]
Other two use cases:

1. **the data to process is not known in advance** - let's say the data to process is not known in advance while I want to keep the application responsive. The data are supplied by an user and he might make a mistake in data and make infinite recursion or infinite loop caused by wrong supplied data (cycles in the trees, non-converging computations, etc), for instance. A typical Vulkan approach would be to submit the work and make a vkWaitForFences() call with UINT64_MAX timeout, which shall result in device lost error after some time. However, it takes even tens of seconds before device lost error is returned. So, this is far from responsive application. Using timeout of 100ms improves the situation to some extent. But after the VK_TIMEOUT return code, GPU is still busy and is going to return device lost soon. It is useless to schedule whatever other work to GPU because device lost is coming. It is not possible to release resources followed by application termination because the resources are still in use by GPU. vkWaitDeviceIdle() or vkWaitQueueIdle() does not help because it just waits tens of seconds until device lost error. It is even dangerous to throw error in C++ code because it might unwind the stack and try to release some Vulkan handles that are still in use. The safe actions seems quite limited. For example, it seems to be safe to (1) print the problem into the console and call std::quick_exit() or (2) create another Vulkan device and start all over, while the current device with all its resources will be released in background thread after device lost error. Having a chance to cancel the job on the queue would allow the first example to gracefully release all resources, and it would prevent device lost error and creation of new device in the second example.
2. **the processing power of the device is not known in advance** - the computation might take too long on particular machine and it might be preferred to return error than to make the user wait. As in the previous case, timeout on fence might be used, but then, it is not possible to clean up resources because they are in use, and will be in use until device lost is generated, which might be much longer than acceptable time period. Examples are the same as in the previous point.

The painful point for me is that after getting VK_TIMEOUT return code, I am lost until device is lost. I cannot do anything, even I cannot throw exception, because I risk that some objects on the stack will be destructed which will try to release some Vulkan handles still in use. One of few meaningful options for responsive application is to leak all Vulkan handles, terminate application and leave up to OS to terminate GPU task and release all the handles.
